cpp/src/utilities/stream_pool.cpp

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/utilities/logger.hpp>
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_pool.hpp>
+
+namespace cudf::detail {
+
+namespace {
+
+// TODO: what is a good number here. what's the penalty for making it larger?
+// Dave Baranec rule of thumb was max_streams_needed * num_concurrent_threads,
+// where num_concurrent_threads was estimated to be 4. so using 32 will allow
+// for 8 streams per thread, which should be plenty (decoding will be up to 4
+// kernels when delta_byte_array decoding is added). rmm::cuda_stream_pool
+// defaults to 16.
+std::size_t constexpr STREAM_POOL_SIZE = 32;
+
+// FIXME: "borrowed" from rmm...remove when this stream pool is moved there
+#ifdef NDEBUG
+#define CUDF_ASSERT_CUDA_SUCCESS(_call) \
+  do {                                  \
+    (_call);                            \
+  } while (0);
+#else
+#define CUDF_ASSERT_CUDA_SUCCESS(_call)                                         \
+  do {                                                                          \
+    cudaError_t const status__ = (_call);                                       \
+    if (status__ != cudaSuccess) {                                              \
+      std::cerr << "CUDA Error detected. " << cudaGetErrorName(status__) << " " \
+                << cudaGetErrorString(status__) << std::endl;                   \
+    }                                                                           \
+    /* NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay) */   \
+    assert(status__ == cudaSuccess);                                            \
+  } while (0)
+#endif
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**💡 suggestion:** ‏ Should be in cuDF's error header?

Reply:
My thinking was it's temporary anyway, but if others feel there's a desire to keep this I'll move it.

Reply:
Is this materially different from `CUDF_CUDA_TRY`? There's an `assert` vs. a `throw`, but I think we _want_ the `CUDF_CUDA_TRY` behavior if I'm not mistaken. https://github.com/rapidsai/cudf/blob/e81d79e94f268499c8656eba1fe8de8122589780/cpp/include/cudf/utilities/error.hpp#L256

Reply:
Ah, is the difference that we need the event destructor calling this to be `noexcept` and thus can't throw? If so, can we mark the destructor `noexcept` explicitly?

Reply:
I think the biggest difference is that the ASSERT version only operates when compiled for debug.  Otherwise it's a no-op.

Reply:
> can we mark the destructor noexcept explicitly?

I'm mildly against this since it's redundant.

Reply:
You’re right. It’s redundant and not necessary. My mistake.

Reply:
Now that I’ve been corrected (thanks), I don’t have strong feelings here. Resolve as you see fit.

Reply:
Ok, I'm going to leave it as is for now. When the stream pool moves to RMM the macro will go away.